Assets/LeapMotion/Core/Scripts/Utils/PostProcessProvider.cs

```diff
@@ -0,0 +1,138 @@
+ * Use subject to the terms of the Leap Motion SDK Agreement available at     *
+ * https://developer.leapmotion.com/sdk_agreement, or another agreement       *
+ * between Leap Motion and you, your company or other organization.           *
+ ******************************************************************************/
+
+using UnityEngine;
+using Leap.Unity.Attributes;
+
+namespace Leap.Unity {
+
+  public abstract class PostProcessProvider : LeapProvider {
+
+    [Tooltip("The LeapProvider whose output hand data will be copied, modified, and "
+           + "output by this post-processing provider.")]
+    [SerializeField]
+    [OnEditorChange("inputLeapProvider")]
+    protected LeapProvider _inputLeapProvider;
+    public LeapProvider inputLeapProvider {
+      get { return _inputLeapProvider; }
+      set {
+        validateInput();
+
+        if (_inputLeapProvider != null) {
+          _inputLeapProvider.OnFixedFrame -= processFixedFrame;
+          _inputLeapProvider.OnUpdateFrame -= processUpdateFrame;
+        }
+
+        _inputLeapProvider = value;
+
+        if (_inputLeapProvider != null) {
+          _inputLeapProvider.OnFixedFrame -= processFixedFrame; // safeguard double-subscription
+          _inputLeapProvider.OnFixedFrame += processFixedFrame;
+          _inputLeapProvider.OnUpdateFrame -= processUpdateFrame; // safeguard double-subscription
+          _inputLeapProvider.OnUpdateFrame += processUpdateFrame;
+        }
+      }
+    }
+
+    public enum DataUpdateMode { UpdateOnly, FixedUpdateOnly, UpdateAndFixedUpdate }
+    [Tooltip("Whether this post-processing provider should process data received from "
+           + "Update frames, FixedUpdate frames, or both. Processing both kinds of frames "
+           + "is only recommended if your post-process is stateless.")]
+    public DataUpdateMode dataUpdateMode = DataUpdateMode.UpdateOnly;
+
+    [Tooltip("When this setting is enabled, frame data is passed from this provider's "
+           + "input directly to its output without performing any post-processing.")]
+    public bool passthroughOnly = false;
+
+    private Frame _cachedUpdateFrame = new Frame();
+    private Frame _cachedFixedFrame = new Frame();
+
+    public override Frame CurrentFrame {
+      get {
+        #if UNITY_EDITOR
+        if (!Application.isPlaying && _inputLeapProvider != null) {
+          processUpdateFrame(_inputLeapProvider.CurrentFrame);
+        }
+        #endif
+        return _cachedUpdateFrame;
+      }
+    }
+
+    public override Frame CurrentFixedFrame {
+      get {
+        #if UNITY_EDITOR
+        if (!Application.isPlaying && _inputLeapProvider != null) {
+          processUpdateFrame(_inputLeapProvider.CurrentFixedFrame);
+        }
+        #endif
+        return _cachedFixedFrame;
+      }
+    }
+
+    protected virtual void OnEnable() {
+      if (_inputLeapProvider == null && Hands.Provider != this) {
+        _inputLeapProvider = Hands.Provider;
```

Comment:
Maybe do a null check here with a nice error message?

Reply:
will take a look

Reply:
I'm not sure an error is warranted here. Sometimes a reference is just null because you're setting up your scene or you have dangling objects for works-in-progress. I modified this check to avoid calling Hands.Provider twice, but otherwise I would propose a silent fail since Hands.Provider isn't fool-proof as it is.